A common interface for all parsers

HawkSK opened the pull request:
This PR makes all generated parsers implement this interface:
```java
package yajco.generator.parsergen;

public interface Parser<T, E extends Exception> {
	T parse(String input) throws E;
	T parse(Reader reader) throws E;
}
```
**Optional:** Make a common superclass ParseException for all thrown exceptions in all parsers (`LisaParseException` , `ParseException`, `LALRParseException`), so the generic parameter can be more specific `E extends ParseException`.


### Use case
A common interface can help to work with the generated parser dynamically. I load the generated class dynamically so I cannot cast it to its static type. With an interface in the `yajco-generator-module` I can cast it to `Parser`, so that I don't need to use reflection.

### Is it a good idea?
I am not sure if it is desired to introduce a dependency on `yajco-generator-module` into the generated files, because also classes like `ParseException`, `SymbolWrapper`, ... are generated and not imported from YAJCo. But what is the benefit of not having a dependency? There is already an import of `yajco.lexer.*` from `yajco-annotations` module.
The modules `yajco-annotation-processor` and all of the `yajco-*-generator-module` already have a dependency on `yajco-generator-module`, which means this is not a new dependency for the user, the user _must_ already have a transitive dependency on `yajco-generator-module`.

sergejx commented:
There is idea that the application should not depend on the parser generator at runtime. Currently, the only module that is needed at runtime is `yajco-annotations`. Over time other classes were added to this module, simply because they were also needed at runtime. First `ReferenceResolver`, later `Lexer`.

The best solution would be, probably, to create a new Maven package called `yajco-runtime`, that would provide things like common lexer implementations and interfaces for generated code.

HawkSK commented:
If the common Parser interface is used as an SPI and during generation the `CompilerGenerator` creates a `yajco.generator.parsergen.Parser` file in `META-INF.services` containing the name of the generated parser, then we could use:

``` java
Path parserJar = Path.of("../examples/yajco-oberon0/target/oberon0-1.0-SNAPSHOT.jar");
var parserClassloader = URLClassLoader.newInstance(new URL[]{parserJar.toUri().toURL()});
Optional<CompilerGenerator> parser = ServiceLoader.load(CompilerGenerator.class, parserClassloader).findFirst();
```

sergejx commented:
@HawkSK, I'm not sure what do you mean by the last comment. Can you provide some example where this would be useful?

HawkSK commented:
@sergejx Sorry I accidentally used `CompilerGenerator` instead of `Parser` in the example. Please see my previous comment updated.

And the use case is to dynamically parse a sentence and access the AST (i.e. domain object model) if you have access to the metamodel (yajco-lang.xml -> Language) and to a sentence of the language.

I personally created a [class ](https://git.kpi.fei.tuke.sk/as421bs/ytm-yajco/-/blob/master/src/main/java/ytm/yajco/translation/language/DynamicParser.java)that loads the parser dynamically. I use it to [dynamically translate any sentence of any YAJCo language into MPS](https://git.kpi.fei.tuke.sk/as421bs/ytm-yajco/-/blob/master/src/main/java/ytm/yajco/translation/SentenceTranslator.java#L59) creating an AST there and registering it as a root node in a model. I aslo used it in my [tests](https://git.kpi.fei.tuke.sk/as421bs/ytm-yajco/-/blob/master/src/test/java/ytm/yajco/translation/GeneratedArtifactsTest.java#L165) to compare the AST of the original sentence with the sentence after importing into MPS and generating some artifacts.

The problem is, that if the `@Parser#className` is not defined, than each CompilerGenerator uses a different naming convention for the generated parser, so you cannot load it dynamically by name easily, I needed to replicate the Beaver default name [manually](url) which strongly couples my solution to the Beaver parser - not good. Having `Parser` as an SPI allows anyone to load the Parser from the compiled jar.

HawkSK commented:
I have also added the common ParseException and implemented the SPI registration to the branch.

However, after trying to move it to a `yajco-runtime` module, I realised that not only the `ReferenceResolver` and `yajco.lexer.*` are needed at runtime, but also beaver itself - transitively through `yajco-beaver-parser-generator-module`. So what should be done:

1. Is it OK to keep the common Parser interface in `yajco-generator-module` because the `yajco-beaver-parser-generator-module` is a runtime dependency and it has this as a transitive dependency anyway?
2. Create a `yajco-*-parser-generator-module-runtime` module for each parser generator separately just to have the runtime dependency explicitly?
3. Something else?

sergejx commented:
I think we can leave the interfaces in `yajco-generator-module`, at least for now. Later we can decide about their separation.
